File: Tl2.2 SOAP_DOTNET_EUREKA_BANK_GR08/02.CLICON/CLICON_EUREKA_GR08/ec.edu.monster.controller/MainController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using CLICON_EUREKA_GR08.LoginServicio;
using CLICON_EUREKA_GR08.CuentaServicio;
using CLICON_EUREKA_GR08.MovimientoServicio;

namespace CLICON_EUREKA_GR08.ec.edu.monster.controller
{
    public class MainController
    {
        private string usuarioActual = null;

        public string GetUsuarioActual()
        {
            return usuarioActual;
        }

        public bool IniciarSesion(string username, string password)
        {
            try
            {
                using (var client = new LoginServicioClient())
                {
                    if (client.Auth(username, password))
                    {
                        usuarioActual = username;
                        return true;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error en la autenticación: " + ex.Message);
            }
            return false;
        }

        public void CerrarSesion()
        {
            usuarioActual = null;
        }

        public bool RealizarDeposito(string cuenta, string monto, string tipo, string cd)
        {
            try
            {
                using (var client = new CuentaServicioClient())
                {
                    return client.Deposito(cuenta, monto, tipo, cd);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error al realizar depósito: " + ex.Message);
                return false;
            }
        }

        public bool RealizarRetiro(string cuenta, string monto)
        {
            return RealizarDeposito(cuenta, monto, "RET", null);
        }

        public bool RealizarTransferencia(string cuentaOrigen, string cuentaDestino, string monto)
        {
            // Seguir la misma lógica del cliente Java: RET + TRA
            bool retiroExitoso = RealizarRetiro(cuentaOrigen, monto);
            if (retiroExitoso)
            {
                return RealizarDeposito(cuentaDestino, monto, "TRA", cuentaOrigen);
            }
            return false;
        }

        public string VerMovimientos(string cuenta)
        {
            try
            {
                List<MovimientoModel> movimientos = null;
                CuentaModel cuentaModel = null;

                using (var movimientoClient = new MovimientoServicioClient())
                using (var cuentaClient = new CuentaServicioClient())
                {
                    movimientos = movimientoClient.Movimientos(cuenta)?.ToList() ?? new List<MovimientoModel>();
                    cuentaModel = cuentaClient.ObtenerCuentaPorNumero(cuenta);
                }

                StringBuilder resultado = new StringBuilder();
                if (movimientos == null || movimientos.Count == 0)
                {
                    resultado.Append("No se encontraron movimientos para la cuenta: ").Append(cuenta).AppendLine();
                    return EncloseInBox("MOVIMIENTOS DE LA CUENTA " + cuenta, resultado.ToString());
                }

                // Mostrar saldo al inicio
                if (cuentaModel != null)
                {
                    resultado.AppendFormat("Saldo Actual de la Cuenta: {0:F2}", cuentaModel.DecCuenSaldo).AppendLine();
                    resultado.AppendLine();
                }

                resultado.AppendFormat("{0,-10} {1,-20} {2,-12} {3,-20} {4,-15} {5,-15} {6,-15}",
                    "Nro", "Fecha", "Tipo Mov.", "Descripción", "Importe", "Cta Ref.", "Saldo").AppendLine();
                resultado.AppendLine("----------------------------------------------------------------------------------------------------");

                foreach (var mov in movimientos)
                {
                    string fecha = mov.FechaMovimientoDt?.ToString("dd/MM/yyyy HH:mm") ?? mov.FechaMovimiento ?? "N/A";
                    string cuentaRef = mov.CuentaReferencia ?? "N/A";
                    string tipoDesc = mov.TipoDescripcion ?? "N/A";

                    resultado.AppendFormat("{0,-10} {1,-20} {2,-12} {3,-20} {4,-15:F2} {5,-15} {6,-15:F2}",
                        mov.NumeroMovimiento,
                        fecha,
                        mov.CodigoTipoMovimiento ?? "N/A",
                        tipoDesc,
                        mov.ImporteMovimiento,
                        cuentaRef,
                        mov.Saldo).AppendLine();
                }

                // Calcular totales
                double totalIngresos = movimientos
                    .Where(m => m.TipoDescripcion != null && !m.TipoDescripcion.Equals("Retiro", StringComparison.OrdinalIgnoreCase))
                    .Sum(m => m.ImporteMovimiento);

                double totalEgresos = movimientos
                    .Where(m => m.TipoDescripcion != null && m.TipoDescripcion.Equals("Retiro", StringComparison.OrdinalIgnoreCase))
                    .Sum(m => m.ImporteMovimiento);

                resultado.AppendLine();
                resultado.AppendLine("RESUMEN:");
                resultado.AppendFormat("Total Ingresos: {0:F2}", totalIngresos).AppendLine();
                resultado.AppendFormat("Total Egresos (Retiros): {0:F2}", Math.Abs(totalEgresos)).AppendLine();
                resultado.AppendFormat("Saldo Neto: {0:F2}", totalIngresos + totalEgresos).AppendLine();

                // Mostrar saldo actual de la cuenta
                if (cuentaModel != null)
                {
                    resultado.AppendFormat("Saldo Actual de la Cuenta: {0:F2}", cuentaModel.DecCuenSaldo).AppendLine();
                }

                return EncloseInBox("MOVIMIENTOS DE LA CUENTA " + cuenta, resultado.ToString());
            }
            catch (Exception ex)
            {
                return "Error al obtener los movimientos: " + ex.Message;
            }
        }

        public string VerDatosCuenta(string cuenta)
        {
            try
            {
                CuentaModel cuentaModel = null;

                using (var client = new CuentaServicioClient())
                {
                    cuentaModel = client.ObtenerCuentaPorNumero(cuenta);
                }

                if (cuentaModel == null)
                {
                    return "No se encontraron datos para la cuenta: " + cuenta;
                }

                StringBuilder resultado = new StringBuilder();
                resultado.AppendFormat("{0,-20} {1,-30}", "Campo", "Valor").AppendLine();
                resultado.AppendLine("--------------------------------------------------");
                resultado.AppendFormat("{0,-20} {1,-30:F2}", "Saldo", cuentaModel.DecCuenSaldo).AppendLine();
                resultado.AppendFormat("{0,-20} {1,-30}", "Contador Movimientos", cuentaModel.IntCuenContMov).AppendLine();
                resultado.AppendLine("--------------------------------------------------");

                return EncloseInBox("DATOS DE LA CUENTA " + cuenta, resultado.ToString());
            }
            catch (Exception ex)
            {
                return "Error al obtener los datos de la cuenta: " + ex.Message;
            }
        }

        // Utilidades para crear recuadros ASCII
        private string EncloseInBox(string title, string multilineContent)
        {
            string[] lines = multilineContent.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
            int maxLineLength = title.Length;
            foreach (string line in lines)
            {
                if (line.Length > maxLineLength)
                {
                    maxLineLength = line.Length;
                }
            }

            int innerWidth = Math.Max(maxLineLength, title.Length);
            string horizontal = Repeat('-', innerWidth + 2);

            StringBuilder box = new StringBuilder();
            box.Append('+').Append(horizontal).Append("+").AppendLine();
            box.Append('|').Append(PadCenter(" " + title + " ", innerWidth + 2)).Append('|').AppendLine();
            box.Append('+').Append(horizontal).Append("+").AppendLine();
            foreach (string line in lines)
            {
                box.Append('|').Append(PadRight(" " + line, innerWidth + 2)).Append('|').AppendLine();
            }
            box.Append('+').Append(horizontal).Append('+');
            return box.ToString();
        }

        private string Repeat(char ch, int count)
        {
            return new string(ch, count);
        }

        private string PadRight(string text, int width)
        {
            if (text.Length >= width)
            {
                return text;
            }
            return text.PadRight(width);
        }

        private string PadCenter(string text, int width)
        {
            if (text.Length >= width)
            {
                return text;
            }
            int totalPadding = width - text.Length;
            int left = totalPadding / 2;
            int right = totalPadding - left;
            return new string(' ', left) + text + new string(' ', right);
        }
    }
}

File: TI2.1  SOAP_JAVA_EUREKA_BANK_GR08/02.CLIESC/CLIESC_EUREKA_GR08/src/main/java/ec/edu/monster/controlador/MainController.java
package ec.edu.monster.controlador;

import ec.edu.monster.service.CuentaService;
import ec.edu.monster.service.LoginService;
import ec.edu.monster.ws.MovimientoModel;
import ec.edu.monster.ws.CuentaModel;
import ec.edu.monster.service.MovimientoService;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.List;

public class MainController {
    private LoginService loginService = new LoginService();
    private CuentaService cuentaService = new CuentaService();
    private MovimientoService movimientoService = new MovimientoService();
    private String usuarioActual = null;

    public String getUsuarioActual() {
        return usuarioActual;
    }

    public boolean iniciarSesion(String username, String password) {
        try {
            String hashedPassword = hashPassword(password);
            if (loginService.auth(username, password)) {
                usuarioActual = username;
                return true;
            }
        } catch (NoSuchAlgorithmException e) {
            System.out.println("Error en la autenticación: " + e.getMessage());
        }
        return false;
    }

    public void cerrarSesion() {
        usuarioActual = null;
    }

    public boolean realizarDeposito(String cuenta, String monto, String tipo, String cd) {
        return cuentaService.realizarDeposito(cuenta, monto, tipo, cd);
    }

    public boolean realizarRetiro(String cuenta, String monto) {
        return cuentaService.realizarDeposito(cuenta, monto, "RET", null);
    }

    public boolean realizarTransferencia(String cuentaOrigen, String cuentaDestino, String monto) {
        boolean retiroExitoso = realizarRetiro(cuentaOrigen, monto);
        if (retiroExitoso) {
            return realizarDeposito(cuentaDestino, monto, "TRA", cuentaOrigen);
        }
        return false;
    }

    public String verMovimientos(String cuenta) {
        try {
            List<MovimientoModel> movimientos = movimientoService.obtenerMovimientos(cuenta);
            CuentaModel cuentaModel = cuentaService.obtenerCuentaPorNumero(cuenta);

            StringBuilder resultado = new StringBuilder();
            if (movimientos == null || movimientos.isEmpty()) {
                resultado.append("No se encontraron movimientos para la cuenta: ").append(cuenta).append("\n");
                
                return encloseInBox("MOVIMIENTOS DE LA CUENTA " + cuenta, resultado.toString());
            }

            // Mostrar saldo al inicio para que no se pierda en listados largos
            if (cuentaModel != null) {
                resultado.append(String.format("Saldo Actual de la Cuenta: %.2f%n", cuentaModel.getDecCuenSaldo()));
                resultado.append("\n");
            }

            resultado.append(String.format("%-10s %-20s %-12s %-20s %-15s %-15s %-15s%n",
                    "Nro", "Fecha", "Tipo Mov.", "Descripción", "Importe", "Cta Ref.", "Saldo"));
            resultado.append("----------------------------------------------------------------------------------------------------\n");

            for (MovimientoModel mov : movimientos) {
                resultado.append(String.format("%-10d %-20s %-12s %-20s %-15.2f %-15s %-15.2f%n",
                        mov.getNumeroMovimiento(),
                        mov.getFechaMovimiento(),
                        mov.getCodigoTipoMovimiento(),
                        mov.getTipoDescripcion(),
                        mov.getImporteMovimiento(),
                        mov.getCuentaReferencia() != null ? mov.getCuentaReferencia() : "N/A",
                        mov.getSaldo()));
            }

            // Calcular totales
            double totalIngresos = movimientos.stream()
                    .filter(m -> !m.getTipoDescripcion().equals("Retiro"))
                    .mapToDouble(MovimientoModel::getImporteMovimiento)
                    .sum();

            double totalEgresos = movimientos.stream()
                    .filter(m -> m.getTipoDescripcion().equals("Retiro"))
                    .mapToDouble(MovimientoModel::getImporteMovimiento)
                    .sum();

            resultado.append("\nRESUMEN:\n");
            resultado.append(String.format("Total Ingresos: %.2f%n", totalIngresos));
            resultado.append(String.format("Total Egresos (Retiros): %.2f%n", Math.abs(totalEgresos)));
            resultado.append(String.format("Saldo Neto: %.2f%n", totalIngresos + totalEgresos));
            
            // Mostrar saldo actual de la cuenta
            if (cuentaModel != null) {
                resultado.append(String.format("Saldo Actual de la Cuenta: %.2f%n", cuentaModel.getDecCuenSaldo()));
            }

            return encloseInBox("MOVIMIENTOS DE LA CUENTA " + cuenta, resultado.toString());
        } catch (Exception e) {
            return "Error al obtener los movimientos: " + e.getMessage();
        }
    }

    public String verDatosCuenta(String cuenta) {
        try {
            CuentaModel cuentaModel = cuentaService.obtenerCuentaPorNumero(cuenta);

            if (cuentaModel == null) {
                return "No se encontraron datos para la cuenta: " + cuenta;
            }

            StringBuilder resultado = new StringBuilder();
            resultado.append(String.format("%-20s %-30s%n", "Campo", "Valor"));
            resultado.append("--------------------------------------------------\n");
            resultado.append(String.format("%-20s %-30.2f%n", "Saldo", cuentaModel.getDecCuenSaldo()));
            resultado.append(String.format("%-20s %-30d%n", "Contador Movimientos", cuentaModel.getIntCuenContMov()));
            resultado.append("--------------------------------------------------\n");

            return encloseInBox("DATOS DE LA CUENTA " + cuenta, resultado.toString());
        } catch (Exception e) {
            return "Error al obtener los datos de la cuenta: " + e.getMessage();
        }
    }

    private String hashPassword(String input) throws NoSuchAlgorithmException {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hashBytes = digest.digest(input.getBytes());
        StringBuilder hexString = new StringBuilder();
        for (byte b : hashBytes) {
            String hex = Integer.toHexString(0xff & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }
        return hexString.toString();
    }

    // Envoltura en recuadro ASCII para mejorar la visualización en consola
    private String encloseInBox(String title, String multilineContent) {
        String[] lines = multilineContent.split("\\r?\\n");
        int maxLineLength = title.length();
        for (String line : lines) {
            if (line.length() > maxLineLength) {
                maxLineLength = line.length();
            }
        }

        int innerWidth = Math.max(maxLineLength, title.length());
        String horizontal = repeat('-', innerWidth + 2);

        StringBuilder box = new StringBuilder();
        box.append('+').append(horizontal).append("+\n");
        box.append('|').append(padCenter(" " + title + " ", innerWidth + 2)).append('|').append("\n");
        box.append('+').append(horizontal).append("+\n");
        for (String line : lines) {
            box.append('|').append(padRight(" " + line, innerWidth + 2)).append('|').append("\n");
        }
        box.append('+').append(horizontal).append('+');
        return box.toString();
    }

    private String repeat(char ch, int count) {
        StringBuilder sb = new StringBuilder(count);
        for (int i = 0; i < count; i++) {
            sb.append(ch);
        }
        return sb.toString();
    }

    private String padRight(String text, int width) {
        if (text.length() >= width) {
            return text;
        }
        StringBuilder sb = new StringBuilder(width);
        sb.append(text);
        while (sb.length() < width) {
            sb.append(' ');
        }
        return sb.toString();
    }

    private String padCenter(String text, int width) {
        if (text.length() >= width) {
            return text;
        }
        int totalPadding = width - text.length();
        int left = totalPadding / 2;
        int right = totalPadding - left;
        StringBuilder sb = new StringBuilder(width);
        for (int i = 0; i < left; i++) sb.append(' ');
        sb.append(text);
        for (int i = 0; i < right; i++) sb.append(' ');
        return sb.toString();
    }
}